srml/offences/src/lib.rs

```diff
@@ -159,8 +161,111 @@ impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T
 			.collect::<Vec<_>>();
 
 		T::OnOffenceHandler::on_offence(
-			&all_offenders,
+			&concurrent_offenders,
 			&slash_perbill,
 		);
+
+		cache.save()
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Compute the ID for the given report properties.
+	///
+	/// The report id depends on the offence kind, time slot and the id of offender.
+	fn report_id<O: Offence<T::IdentificationTuple>>(
+		time_slot: &O::TimeSlot,
+		offender: &T::IdentificationTuple,
+	) -> ReportIdOf<T> {
+		(O::ID, time_slot.encode(), offender).using_encoded(T::Hashing::hash)
+	}
+
+	/// Goes through all offenders in a report and record each unique offence updating indexes.
+	///
+	/// Returns `true` if there were unique reports.
+	fn collect_unique_reports<O: Offence<T::IdentificationTuple>>(
```

Comment:
I think I'd rather rename it so something like: `store_new_offenders` or at least `collect_new/unique_offenders`